### ISSUE TITLE
OnDelete updateStrategy of mysql statefulset

### ISSUE THREAD

**Issue by jianhaiqing** [presslabs/mysql-operator#562]
I’m wondering whether it would be better if the updateStrategy OnDelete of statefulset is used. If that, we can’t control the scale-down process, graceful-master-takeover before delete the pod if it’s master.
so far, RollingUpdate updateStrategy is used.

**Comment by AMecea** [comment]
Hi @jianhaiqing, yes this a wanted feature :smile: but we didn't have the time to test and implement it. (please see also #404 )

I will love to see this implemented. Let's discuss this. what are your ideas about this, how do you see it implemented?

**Comment by jianhaiqing** [comment]
There're the solutions I'm thinking over.
S-1. Add preStop in mysql container: preStop will find out whether the current pod is master or not, if that failover will be taken
S-2: If the replicas of statefulset is bigger than that of mysqlcluster, scale-downing is in-progress, operator will check if the max-pod is master or not, if that failover will be taken before setting the size of statefulset.

[graceful-master-takeover-auto api will be used in this situation, orchetrator should be upgrader to lastest version](https://github.com/openark/orchestrator/blob/d2198443d9606373033daa46e6c18fb724351b18/go/http/api.go#L3099)

**Comment by jianhaiqing** [comment]
Another feature should be considered, if the resource of statefulset is changed, each pod should be deleted one by one, so the request resource of each pod can take effect after its reconstruction.

**Comment by AMecea** [comment]
Those solutions sound good but there are some related issues that need to be addressed. This feature can become very complicated so let's split it into two, as you already did by dividing into two solutions.

First, using **preStop hook** (S-1) it's a good idea and I think that this is not optional (and not mutually exclusive to S-2) and we should start from here. A possible implementation can be:
 * create an endpoint in `sidecar` container which will listen for pre stop event and will connect to Orchestrator and trigger `graceful-master-takeover-auto`. The issue here is the connection to Orchestrator, which should be protected. So you have to authenticate to Orchestrator either by implementing a proxy server into MySQL Operator, which is better or by directly insert the credentials into the pod, which is faster. 
 * make sure that `sidecar` container dies after `mysql` container. This can be done with a preStop hook too.
 * we should NOT build our custom docker images for MySQL. We want to be compatible with the official ones.

This will be useful in many scenarios like evicted nodes, manually stopped containers, scale down, etc.

The second step is to use **the `OnDelete` update strategy** for statefulset this will be helpful when doing updates because the operator can choose in which order the update happens and by doing this it will prevent a double failover.

What do you think? Will this cover all your use cases? Did I miss something?

**Comment by jianhaiqing** [comment]
I agree with u, these two solutions are not exclusive. 
"create an endpoint in sidecar container which will listen for pre stop event", i don't know how to implement the feature. Can u elaborate it? 
As far as I'm concerned, i would add an shell script as configmap to the preStop of MySQL container. 
The script will check if the current pod is master or not, if asserted, api graceful-takeover-auto of orchestrator will be curl.

**Comment by jianhaiqing** [comment]
@AMecea 
I have implement OnDeleteStatefulSetStrategyType, it looks like the follow code.
Can you help me review the solution?
```golang
func (s *sfsSyncer) SyncFn(ojb runtime.Object) error {
	sst := ojb.(*apps.StatefulSet)

	s.cluster.Status.ReadyNodes = int(sst.Status.ReadyReplicas)
	sst.Spec.Replicas = s.cluster.Spec.Replicas

	isScaleDown, err := s.checkIfScaleDownInProgress(sst)
	if err != nil || isScaleDown {
		return err
	}
       // check cluster resource, select-label and affinity is different from statefulset.
	isUpdateSst, reason, msg := s.shouldManuallyUpdateStatefulset(sst)
	if err != nil {
		return err
	}
	if isUpdateSst {
		// update statefulset
		// TODO: build statefulset
		s.buildStatefulsetSpec(sst)
		if err := s.client.Update(context.TODO(), sst); err != nil {
			return err
		}
		return s.cluster.CreateEventForCluster(s.client, s.scheme, reason, msg)
	}

	if err := s.checkUpgradePod(sst); err != nil {
		return err
	}

	s.buildStatefulsetSpec(sst)
	return nil
}
func (s *sfsSyncer) checkUpgradePod(sst *apps.StatefulSet) error {
	if sst.Status.CurrentRevision == sst.Status.UpdateRevision {
		return nil
	}
	if sst.Status.Replicas == sst.Status.UpdatedReplicas {
		return nil
	}
	// check if pod is all ready
	podList, err := s.GetStatefulSetPods(sst, s.cluster.Namespace)
	if err != nil {
		return err
	}
	waitingUpgradeReplicaPods := []core.Pod{}
	waitingUpgradeMasterPods := []core.Pod{}

	var readyPods int32
	readyPods = 0
	for _, pod := range podList.Items {
		if podutils.IsPodReady(&pod) {
			readyPods++
		}
		revision, exist := pod.ObjectMeta.Labels[apps.ControllerRevisionHashLabelKey]
		if !exist {
			return errors.New(fmt.Sprintf("[upgrade] mysqlcluster: [%s/%s]'s pod: [%s] has no label: %s", s.cluster.Namespace, s.cluster.Name, pod.Name, apps.ControllerRevisionHashLabelKey))
		}
		podRole, exist := pod.Labels[wrapper.RoleLabel]
		if !exist {
			return errors.New(fmt.Sprintf("[upgrade] mysqlcluster: [%s/%s]'s pod: [%s] has no label: %s", s.cluster.Namespace, s.cluster.Name, pod.Name, wrapper.RoleLabel))

		}
		if revision != sst.Status.UpdateRevision {
			switch podRole {
			case string(wrapper.MasterRole):
				waitingUpgradeMasterPods = append(waitingUpgradeMasterPods, pod)
			case string(wrapper.ReplicaRole):
				waitingUpgradeReplicaPods = append(waitingUpgradeReplicaPods, pod)
			}
		}
	}
	if readyPods != *s.cluster.Spec.Replicas {
		return errors.New(fmt.Sprintf("[upgrade]ready  pods number is  [%d], totalPod number is %d",
			readyPods, *s.cluster.Spec.Replicas))
	}

	if len(waitingUpgradeReplicaPods) > 0 {
		pod := waitingUpgradeReplicaPods[0]
		log.Info(fmt.Sprintf("[upgrade]waitingUpgradeReplicaPods=%d, start to delete pod %s/%s",
			len(waitingUpgradeReplicaPods), pod.Namespace, pod.Name), "deletepod", pod)

		if err := s.client.Delete(context.TODO(), &pod); err != nil {
			return err
		}
	} else if len(waitingUpgradeMasterPods) > 0 {
		pod := waitingUpgradeMasterPods[0]
		log.Info(fmt.Sprintf("[upgrade]waitingUpgradeMasterPods=%d, start to failover before delete pod %s/%s",
			len(waitingUpgradeReplicaPods), pod.Namespace, pod.Name), "deletepod", pod)
		err := s.gracefulMasterTakeoverAuto("mysql-operator", "Upgrade",
			fmt.Sprintf("masterpod=%s, currentpod=%s/%s", s.cluster.GetMasterHost(), pod.Namespace, pod.Name))
		if err != nil {
			return err
		} else {
			return errors.New("[upgrade]gracefulMasterTakeoverAuto is done before upgrade master")
		}
	}

	return nil
}
```

**Comment by jianhaiqing** [comment]
One more thing, orc-helper, Since we can update resource , select-labels , and do graceful-master-takeover-auto before kill the pod; The question is, do we need to kill the old connections of old-master in orc-helper, which is used in orchestrator hook.

**Comment by mhumpula** [comment]
Just a quick idea: what about using pod finalizer to trigger clean failover in orchestrator? Then the statefulset controller can restart pods as it likes. I'm currently working on readiness gate for pods to signalize that freshly started replica is not only MySQL ready, but it actually caught up with master.

**Comment by AMecea** [comment]
@mhumpula this is not possible because kubelet will kill containers even if they have finalizers set, see https://github.com/kubernetes/kubernetes/issues/73584

@jianhaiqing, the logic is good but I think will be better to implement this in a different controller. We already have [a node controller](https://github.com/presslabs/mysql-operator/blob/master/pkg/controller/node/node_controller.go#L169) that reconciles a pod. You can extract [the current logic](https://github.com/presslabs/mysql-operator/blob/master/pkg/controller/node/node_controller.go#L214-L257) in a function and call your function that does the upgrade too.  I'm saying this because the statefulset syncer should be used only one purpose (that is of handling sts). We already have some [pod syncers](https://github.com/presslabs/mysql-operator/blob/master/pkg/controller/mysqlcluster/mysqlcluster_controller.go#L260-L268) that operates on the pod. Doing this from too many places will be hard to follow and can be buggy (I plan to move that logic into node controller too).

The problem with this approach is that in that controller you can't iterate over pods so you have to use another algorithm. You can determine if the current pod is master or not, also you can query for its related statefulset (or store on the cluster status current pod revision) and you can determine if all pods are ready or not also you can determine how many pods are updated. Using this the algorithm can be:
```
if ( ! all-pods-are-ready ) return nil

if ( master && every-thing-eles-is-updated)
  delete_pod

if ( replica)
  delete_pod
```

What do you think?

**Comment by jianhaiqing** [comment]
I have tested the current implementation a lot, it works good. I plan to deploy the current version in our test environment. We can do  refactor in next stage. 
Can I merge the current implementation, then do refactoring ?

There are two questions i'm not that clear. 
1. About preStop, I use a configmap to add a shell in MySQL container as preStop.  what do you mean by "create an endpoint in sidecar container which will listen for pre stop event" ？

2. unfortunately, i forgot the second one.

**Comment by AMecea** [comment]
Yes, please open a PR. My question is how this approach is better than the default update strategy (statefulset rolling update) with preStop hook because it's almost the same?

Your idea of mounting the script it's ok, no need to add an endpoint on the sidecar container. What I suggest is good in the case when the script file is too complicated.

**Comment by jianhaiqing** [comment]
Out of curiosity, how could you implement an endpoint on the sidecar container to listen the preStop event, I'm not that familiar with k8s. I really appreciate if you can explain the implementation.

With OnDelete update strategy, you can delete  all pods in order, you can select all the replicas, and upgrade all the replicas and upgrade the master at last.
And with RollingUpdate strategy, and the worst situation might occur like this, 
```
Pod_N-Master
  + pod_(N-1)-replica
  + pod_(N-2)-replica
```
Failover might happen when each pod is deleted during rolling update. And every failover takes at least 3 seconds in orchestrator exclude the endpoint switch of service.

**Comment by AMecea** [comment]
The sidecar container starts [an HTTP server](https://github.com/presslabs/mysql-operator/blob/master/pkg/sidecar/server.go#L54) which expose backups. We can extend this server to have an endpoint that can be called from `mysql` container using [HTTPGetAction](https://godoc.org/k8s.io/api/core/v1#HTTPGetAction) which will the same thing as the script.

I agree with you, but what I want to highlight here is that it's more important to have implemented preStop hooks before OnDelete policy because you can simplify the logic, you won't need to add `gracefulMasterTakeoverAuto` before master upgrade because lifecycle hooks will do the rest. Moreover, maybe we don't need at all OnDelete because preStop hooks may be enough.

Anyway, I appreciate your good work, keep doing it!

**Comment by jianhaiqing** [comment]
If someone is going to use the operator in product environment, i suppose he probably wanna OnDelete, then only one failover might happen during rolling update.

**Comment by jianhaiqing** [comment]
preStop is added in following PR
https://github.com/presslabs/mysql-operator/pull/576

**Comment by stankevich** [comment]
We're running a slightly modified version of the operator where `podSpec` accepts any custom `lifecycle` configuration and passes it to `mysql` container. This allows us to execute a `preStop` script that checks if the current pod is the master and, if so, gracefully moves the master away by `curl`ing the Orchestrator API via `mysql-operator` service (no auth required). The script can check the pod index and move the master role to a designated pod to avoid a double failover. This can still result in two graceful recoveries if the pod with the highest index is the master, but the overall mechanism is really simple and it might be a good alternative to the `OnDelete` strategy. (We also immediately auto-acknowledge all successful recoveries in Orchestrator to unblock any new ones.)

That said, I'd like to suggest making the StatefulSet's `updateStrategy` configurable. Perhaps `OnDelete` could become the new default, but there should still be an option to use `RollingUpdate` if needed. This would also mean making `podSpec.lifecycle` configurable.

**Comment by AMecea** [comment]
@stankevich would you like to share the pre stop script? 

>  (We also immediately auto-acknowledge all successful recoveries in Orchestrator to unblock any new ones.)

I think we can do it too, we can make [this variable](https://github.com/presslabs/mysql-operator/blob/4d88be22e4b30c65c1a0fd60794533cbd4fe50e7/pkg/controller/orchestrator/orchestrator_reconcile.go#L42) configurable from command line.
